rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,234 @@
+
+### Implementation Details
+
+This section is divided in the following subsections:
+- [Data tracked](#data-tracked): which data about each agent will be tracked
+- [Data collection](#data-collection): how such data will flow from the agents to PreHog
+- [Data computation](#data-computation): how to compute such data
+
+#### Data tracked
+
+We want to start tracking the following data in PreHog:
+
+1. Teleport version
+2. Teleport access protocols (`ssh`, `kube`, `app`, `db` and `windows_desktop`)
+3. OS (`linux` or `darwin`, as these are the only two OS currently supported)
+4. OS version (e.g. Linux distribution)
+5. Host architecture (e.g. `amd64`)
+6. `glibc` version
+7. [Installation methods](https://goteleport.com/docs/installation/) (Dockerfile, Helm, `install-node.sh` and `*-ad*.ps1` scripts)
+8. Container runtime (e.g. Docker)
+9. Container orchestrator (e.g. Kubernetes)
+10. Cloud environment (e.g. AWS, GCP, Azure)
+
+#### Data collection
+
+Currently, when an agent first starts, the [inventory control system](https://github.com/gravitational/teleport/tree/6f9ad9553a5b5946f57cb35411c598754d3f926b/lib/inventory) sends an [`UpstreamInventoryHello`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/api/proto/teleport/legacy/client/proto/authservice.proto#L1936-L1953) message to the auth server.
+This message has the following fields:
+
+```protobuf
+message UpstreamInventoryHello {
+  string Version = 1;
+  string ServerID = 2;
+  repeated string Services = 3 [(gogoproto.casttype) = "github.com/gravitational/teleport/api/types.SystemRole"];
+  string Hostname = 4;
+}
+```
+
+The `Version` field contains the Teleport version, while the `Services` field contains the subset of the system roles that are currently active at the agent.
+
+We will extend this message to contain the remaining agent data that we want to track:
+
+```protobuf
+message UpstreamInventoryHello {
+  // (...)
+  string OS = 5;
+  string OSVersion = 6;
+  string HostArchitecture = 7;
+  string GLibCVersion = 8;
+  repeated string InstallMethods = 9;
+  string ContainerRuntime = 10;
+  string ContainerOrchestrator = 11;
+  string CloudEnvironment = 12;
+}
+```
+
+When the auth server receives an `UpstreamInventoryHello` message, it will take the information in the message and send it to PreHog.
+For this, a new PreHog `AgentMetadataEvent` message will be added (note that only the `UpstreamInventoryHello.Hostname` won't be sent to PreHog as it can contain PII but also because it doesn't seem useful):
+
+```protobuf
+message AgentMetadataEvent {
+  string version = 1;
+  string server_id = 2;
+  repeated TeleportAccessProtocol protocols = 3;
+  string os = 4;
+  string os_version = 5;
+  string host_architecture = 6;
+  string glibc_version = 7;
+  repeated string install_methods = 8;
+  string container_runtime = 9;
+  string container_orchestrator = 10;
+  string cloud_environment = 11;
```

Comment:
How will AgentMetadataEvent be tied back to an account?

Reply:
IIUC before sending events to PreHog, we wrap them in a [`SubmitEventRequest`](https://github.com/gravitational/teleport/blob/f622510dd08944d8176f69a102cf8991d0f62e84/proto/prehog/v1alpha/teleport.proto#L300) which has a `ClusterName` that is set here: https://github.com/gravitational/teleport/blob/8dfa07559bd2b13ce470747d99ea551b0ff81380/lib/services/usagereporter.go#L577